lint/rule.go

```diff
@@ -14,10 +14,295 @@
-	// Info returns metadata about a rule.
-	Info() RuleInfo
-	// Lint performs the linting process.
-	Lint(Request) ([]Problem, error)
+	// GetName returns the name of the rule.
+	GetName() RuleName
+
+	// GetURI returns the URI where the applicable guideline
+	// is documented. (This should generally be an AIP on https://aip.dev/.)
+	GetURI() string
+
+	// Lint accepts a FileDescriptor and lints it,
+	// returning a slice of Problem objects it finds.
+	Lint(*desc.FileDescriptor) []Problem
+}
+
+// FileRule defines a lint rule that checks a file as a whole.
+type FileRule struct {
+	Name RuleName
+	URI  string
+
+	// LintFile accepts a FileDescriptor and lints it, returning a slice of
+	// Problems it finds.
+	LintFile func(*desc.FileDescriptor) []Problem
+
+	noPositional struct{}
+}
+
+// GetName returns the name of the rule.
+func (r *FileRule) GetName() RuleName {
+	return r.Name
+}
+
+// GetURI returns the URI where the applicable guideline is documented.
+func (r *FileRule) GetURI() string {
+	return r.URI
+}
+
+// Lint forwards the FileDescriptor to the LintFile method defined on the
+// FileRule.
+func (r *FileRule) Lint(fd *desc.FileDescriptor) []Problem {
+	return r.LintFile(fd)
+}
+
+// MessageRule defines a lint rule that is run on each message (top-level or
+// nested) within a file.
+type MessageRule struct {
+	Name RuleName
+	URI  string
+
+	// LintMessage accepts a MessageDescriptor and lints it, returning a slice
+	// of Problems it finds.
+	LintMessage func(*desc.MessageDescriptor) []Problem
+
+	noPositional struct{}
+}
+
+// GetName returns the name of the rule.
+func (r *MessageRule) GetName() RuleName {
+	return r.Name
+}
+
+// GetURI returns the URI where the applicable guideline is documented.
+func (r *MessageRule) GetURI() string {
+	return r.URI
+}
+
+// Lint accepts a FileDescriptor and iterates over every message in the
+// file, and lints each message in the file.
+func (r *MessageRule) Lint(fd *desc.FileDescriptor) []Problem {
+	problems := []Problem{}
+
+	// Iterate over each message and process rules for each message.
+	for _, message := range getAllMessages(fd) {
+		problems = append(problems, r.LintMessage(message)...)
+	}
+	return problems
+}
+
+// FieldRule defines a lint rule that is run on each field within a file.
+type FieldRule struct {
+	Name RuleName
+	URI  string
+
+	// LintField accepts a FieldDescriptor and lints it, returning a slice of
+	// Problems it finds.
+	LintField func(*desc.FieldDescriptor) []Problem
+
+	noPositional struct{}
+}
+
+// GetName returns the name of the rule.
+func (r *FieldRule) GetName() RuleName {
+	return r.Name
+}
+
+// GetURI returns the URI where the applicable guideline is documented.
+func (r *FieldRule) GetURI() string {
+	return r.URI
+}
+
+// Lint accepts a FileDescriptor and lints every field in the file.
+func (r *FieldRule) Lint(fd *desc.FileDescriptor) []Problem {
+	problems := []Problem{}
+
+	// Iterate over each message and process rules for each field in that
```

Comment:
What about extensions?

Reply:
What about them?

Reply:
Will an extension be considered as a field?

Reply:
No. Among many other reasons, it would be unlikely to be in the same file.